scipy/optimize/tests/test_nonlin.py

```diff
@@ -168,6 +171,7 @@ def _check_root(self, f, method, f_tol=1e-2):
     def _check_func_fail(self, *a, **kw):
         pass
 
+    @pytest.mark.filterwarnings('ignore::DeprecationWarning')
```

Comment:
This patch removes the need to filter the deprecation warnings. It looks like there was a minor oversight in #8400 and how the tolerance is passed to the iterative solver as currently as far as I can tell the old `if 'tol' in self.method_kw:` branch is never hit. However, with this patch there is now one test that overflows :confused: 
```
diff --git a/scipy/optimize/_nonlin.py b/scipy/optimize/_nonlin.py
index c6abe1684..2812d01e0 100644
--- a/scipy/optimize/_nonlin.py
+++ b/scipy/optimize/_nonlin.py
@@ -1477,7 +1477,7 @@ class KrylovJacobian(Jacobian):
         return r
 
     def solve(self, rhs, tol=0):
-        if 'tol' in self.method_kw:
+        if 'atol' in self.method_kw:
             sol, info = self.method(self.op, rhs, **self.method_kw)
         else:
             sol, info = self.method(self.op, rhs, tol=tol, **self.method_kw)
diff --git a/scipy/optimize/tests/test_nonlin.py b/scipy/optimize/tests/test_nonlin.py
index d60665b6b..54100020a 100644
--- a/scipy/optimize/tests/test_nonlin.py
+++ b/scipy/optimize/tests/test_nonlin.py
@@ -171,7 +171,6 @@ class TestNonlin:
     def _check_func_fail(self, *a, **kw):
         pass
 
-    @pytest.mark.filterwarnings('ignore::DeprecationWarning')
     def test_problem_nonlin(self):
         for f in [F, F2, F2_lucky, F3, F4_powell, F5, F6]:
             for func in SOLVERS.values():
@@ -181,7 +180,6 @@ class TestNonlin:
                     continue
                 self._check_nonlin_func(f, func)
 
-    @pytest.mark.filterwarnings('ignore::DeprecationWarning')
     @pytest.mark.parametrize("method", ['lgmres', 'gmres', 'bicgstab', 'cgs',
                                         'minres', 'tfqmr'])
     def test_tol_norm_called(self, method):
@@ -197,7 +195,6 @@ class TestNonlin:
                              tol_norm=local_norm_func)
         assert_(self._tol_norm_used)
 
-    @pytest.mark.filterwarnings('ignore::DeprecationWarning')
     def test_problem_root(self):
         for f in [F, F2, F2_lucky, F3, F4_powell, F5, F6]:
             for meth in SOLVERS:
```

Reply:
Hmm, I am not sure I understand your point though. Do you mind elaborating? Is this related to ignoring the warning or something wrong with the test?

Reply:
Ah OK. Got it. Hopefully we can convert all solvers to accept atol, rtol before 1.12 is out hence we can unify all solvers under one common API. That `*` business is really a good idea in the signature to enforce it.

Reply:
+1 sounds like a good plan

Reply:
Sidenote: In my notifications, I got a long comment from Jake here that is now gone. I'm not sure if it was deleted (cannot find it in the edits), but that makes things hard to follow. It's unavoidable to edit comments occasionally, but can we avoid deleting them?

Reply:
Sorry, I wrote that comment and then realised I had miss understood part of the code in `scipy/optimize/_nonlin.py` and that the comment was going to cause unnecessary confusion so deleted it.